Public/Src/Cache/ContentStore/Library/FileSystem/PassThroughFileSystem.cs

```diff
@@ -445,89 +446,86 @@ private async Task CopyFileInsideSemaphoreAsync(AbsolutePath sourcePath, Absolut
+                // Trying to change the attributes and delete the file.
 
-                int lastError = Marshal.GetLastWin32Error();
-                if (handle.IsInvalid && mode == FileMode.Create)
+                if (exception.NativeErrorCode() == ERROR_ACCESS_DENIED &&
+                    (GetFileAttributes(path) & FileAttributes.ReadOnly) != 0)
                 {
-                    switch (lastError)
-                    {
-                        case NativeMethods.ERROR_ACCESS_DENIED:
-                        case NativeMethods.ERROR_SHARING_VIOLATION:
-                            if (lastError == NativeMethods.ERROR_ACCESS_DENIED &&
-                                (GetFileAttributes(path) & FileAttributes.ReadOnly) != 0)
-                            {
-                                SetFileAttributes(path, FileAttributes.Normal);
-                            }
-                            
-                            DeleteFile(path);
-                            openHandle();
-                            break;
-                    }
+                    SetFileAttributes(path, FileAttributes.Normal);
                 }
 
-                if (handle.IsInvalid)
+                DeleteFile(path);
+                (result, exception) = tryOpenFile();
+            }
+
+            if (exception != null)
+            {
+                switch (exception.NativeErrorCode())
                 {
-                    switch (lastError)
-                    {
-                        case NativeMethods.ERROR_FILE_NOT_FOUND:
-                        case NativeMethods.ERROR_PATH_NOT_FOUND:
-                            return (Stream)null;
-                        default:
-                            throw ThrowLastWin32Error(
+                    case ERROR_FILE_NOT_FOUND:
+                    case ERROR_PATH_NOT_FOUND:
+                        return (Stream)null;
+                    default:
+                        throw ThrowLastWin32Error(
+                            path.Path,
+                            string.Format(
+                                CultureInfo.InvariantCulture,
+                                "Unable to open a file '{0}' as {1} with {2}.",
                                 path.Path,
-                                string.Format(
-                                    CultureInfo.InvariantCulture,
-                                    "Unable to open a handle to '{0}' as {1} with {2}.",
-                                    path.Path,
-                                    mode,
-                                    accessMode));
-                    }
+                                mode,
+                                accessMode),
+                            exception.NativeErrorCode());
                 }
+            }
 
-                var needToDisposeHandle = true;
+            return result;
+
+            (FileStream fileStream, Exception Exception) tryOpenFile()

```

Comment:
>(FileStr [](start = 12, length = 8)

Would Either<T,U> make more sense?

Reply:
yep, but without proper union types this is effectively the same thing.